[Edits] 2022-02-08-Junior-engineers:-your-job-is-to-learn.md (#3)
alexturek/alexturek.github.io: _posts/2022-02-08-Junior-engineers:-your-job-is-to-learn.md

An extremely common misconception among junior engineers is that they should be as productive as senior engineers and, at the end of the day, the people who suffer the most because of it, are the junior engineers themselves.

# A Career of Learning

> The second most common thing to do with technology is have to learn some new tech stack or language. The first - I hope! - is building, or dreaming of building something, something valuable to the world with it BUT you can't do that until you've learned at least one tech well. _(This post doesn't even address the hard part nobody talks about: how to work _on a team_ to build software. Another day.)_

Reviewer: ```suggestion
The second most common thing to do with technology is have to learn some new tech stack or language. The first - I hope! - is building, or dreaming of building something, something valuable to the world. But you can't do that until you've learned at least one tech well. _(This post doesn't even address the hard part nobody talks about: how to work on a team to build software. Another day.)_
```

> Just like everything in life, learning is easier with practice. After ~15 years in tech, learning new technologies has become simple for me - there's a ton of pattern recognition and instinctive navigation of documentation that come more naturally after seeing it in different iterations so many times. Sometimes it's even a pleasant little surprise!

Reviewer: ```suggestion
Just like everything in life, learning is easier with practice. After ~15 years in tech, learning new technologies has become simple for me - there's a ton of pattern recognition and instinctive navigation of docs. It comes naturally after seeing it in different iterations so many times. Sometimes it's even a pleasant little surprise!
```

> _Oh, I get to learn Kubernetes finally! Time to write some YAML!_
>
> (This is when the pleasantness fades)

> But early on in your career, learning can be awful. You struggle for days trying to do something that takes more senior engineers 5 minutes to solve. You get misled by StackOverflow posts, or can't figure out the right search term, or wander into a completely new concept (_"what is a cluster??"_) that you never learned before, and suddenly have to stop building and start learning from 0. Some weeks, you go down 10 dead ends without actually figuring out your problem and that's understandably frustrating!

Reviewer: ```suggestion
But early on in your career, learning can be awful. You struggle for days trying to do something that takes more senior engineers 5 minutes to solve. You get misled by StackOverflow posts, or can't figure out the right search term, or wander into a completely new concept (_"what is a cluster??"_) that you never learned before, and suddenly have to stop building and start learning from 0. Some weeks, you go down 10 dead ends without actually figuring out your problem. I went through it and I remember how frustrating that is!
```

# When You Don't Learn

I see junior engineers get sidelined from that good, hard stuff, and they often have two failure modes:

1. ## "Oh No! I'm Stuck" 
   ### **So focused on beating themselves up that they aren't learning.**

    This is understandable, and I remember my first job and the wild emotional swings I went through when I struggled for days or weeks to overcome some seemingly simple problem.

    Talk to your mentor or a more senior engineer, be honest with them, ask for feedback. 9 times out of 10, you're doing fine.

>     Or, make sure you're writing lessons down. Keep a scratchpad of "Today I Learned." _Here's how I split an array in JavaScript! Here's how I search a git repo for a string!_ Set a reminder at a monthly or quarterly interval and revisit that file: seeing that whole list you will stop feeling so much imposter syndrome.

Reviewer: ```suggestion
    Or, make sure you're writing lessons down. Keep a scratchpad of "Today I Learned." _Here's how I split an array in JavaScript! Here's how I search a git repo for a string!_ Set a reminder at a monthly or quarterly interval and revisit that file: seeing that whole list will help you push back your imposter syndrome.
```

>     When I see somebody, for example, panicking about not updating sprint tasks fast enough, I try to take them aside and ask what they've been learning. Focusing on your success helps you realize it exists.

Reviewer: ```suggestion
    **Senior engineers**: When I see somebody, for example, panicking about not updating sprint tasks fast enough, I try to take them aside and ask what they've been learning. Focusing on their success helps them realize it exists.
```

    
2. ## "Fake It Till You Make It" 
   ### Rejecting feedback from more experienced people trying to help them learn.**

    You can and will learn plenty on your own. You will also learn more, not less, from other people as you progress in your career.

    An immature person starts a disagreement with a zero-sum mentality, assuming that if someone has a better suggestion, that means they "lose" something. If you catch yourself working hard to defend your technical choices, or designs, or code, ask some questions.

    * To them: _How'd you learn this? What's an example of X being better than Y in the real world?_
    * To yourself: _What do I know that this person doesn't? What do they know that I don't? Can I find at least one person who agrees with me?_

>     This failure mode can be painful and frustrating. It teaches me patience, and then how to communicate better, and then finally that maybe I'm taking my job too seriously and I should get a drink at 2pm.

Reviewer: ```suggestion
    **Senior engineers**: This failure mode can be painful and frustrating. It teaches me patience, and then how to communicate better, and then finally that maybe I'm taking my job too seriously and I should get a drink at 2pm.
```

    When I encounter one of these people, I usually treat it as an expensive (for the rest of the team) but necessary learning exercise for everybody. They're going to make a mess, they're going to ship some garbage code, and I'm not going to stop it. Usually the right strategy is containment. Give them a playground to "prove" their ideas out and in 6 months or a year, they and the entire team will be struggling with their work. It sucks that you have to maintain it, but hopefully nobody on that team will ever repeat that type of mistake.

# Embrace The Hard Stuff

> Your job early on isn't to magically avoid all the First-Time Learnies (TM). It's to go through them, struggle, and learn a ton of crap that seems useless but you'll use later. Right now you're building a huge amount of context, and how to apply it. The next time there'll be far less context to build, and learning will feel easier (not easy! not yet! but easier). And after you've done this a few times you get amazingly good at pattern recognition. You've learned technologies that are similar enough to understand the new one.

Reviewer: 👍
